<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>BEAST2 to TransPhylo to Regression Pipeline • beast2tpPipeline</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="BEAST2 to TransPhylo to Regression Pipeline">
<meta name="description" content="Given fasta files and corresponding metadata, this package goes through the process of splitting fasta files into SNP clusters, creating an XML file for BEAST2 (constrained to a particular model - use BEAUti otherwise), running BEAST2, running TransPhylo, and then doing some kind of regression with the TransPhylo results. This package was developed with the specific application of the Kopanyo project with TB data, so some of it is highly specific and not very transferable.">
<meta property="og:description" content="Given fasta files and corresponding metadata, this package goes through the process of splitting fasta files into SNP clusters, creating an XML file for BEAST2 (constrained to a particular model - use BEAUti otherwise), running BEAST2, running TransPhylo, and then doing some kind of regression with the TransPhylo results. This package was developed with the specific application of the Kopanyo project with TB data, so some of it is highly specific and not very transferable.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">beast2tpPipeline</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html" aria-label="Home"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/mcc_tree_pipeline.html">Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jessalynnsebastian/beast2tpPipeline" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="beast2tppipeline-package">beast2tpPipeline Package<a class="anchor" aria-label="anchor" href="#beast2tppipeline-package"></a>
</h1></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package was created to work with data from the Kopanyo study, namely fasta files of TB sequences and associated metadata. It defines a clear pipeline from fasta files of sequences and associated metadata, through creating SNP clusters, running BEAST2, running TransPhylo, and doing regression on the results.</p>
<p>In <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010696" class="external-link">Using genetic data to identify transmission risk factors: Statistical assessment and application to tuberculosis transmission</a>, Goldstein et al. show that this pipeline is very biased; the BEAST2-TransPhylo pipeline has low sensitivity of identifying infection sources and biases regression coefficients toward zero. This package was built to facilitate exploration of the qualities of this pipeline.</p>
<p>This package is built such that each step in the pipeline has an associated function. The steps and functions are:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Assigning SNP clusters</strong>: <code>assign_snp_clusters</code> takes a set of sequences and metadata, and assigns sequences to SNP clusters.</li>
<li>
<strong>Creating BEAST2 clusters</strong>: <code>create_BEAST2_clusters</code> takes the SNP clusters and separates the sequence data by cluster, keeping only the SNPs if desired.</li>
<li>
<strong>Creating XML files for BEAST2</strong>: <code>create_cluster_xml</code> takes the SNP data and creates an XML file for BEAST2 to run.</li>
<li>
<strong>Running BEAST2</strong>: <code>run_beast2</code> runs BEAST2 on the XML file. This is where we take the DNA sequences and infer a timed phylogenetic tree, which tells us about the ancestry of the pathogen.</li>
<li>
<strong>Getting the MCC trees</strong>: <code>get_mcctree</code> takes the posterior trees from BEAST2 and creates a maximum clade credibility tree - a summary tree, like a mean or median. (Alternatively, <strong>Getting a Sample of Posterior Trees</strong>: <code>sample_BEAST2_trees</code> takes the posterior trees from BEAST2 and subsamples a specified number of them.)</li>
<li>
<strong>Running TransPhylo on the MCC trees</strong>: <code>run_TransPhylo</code> runs TransPhylo on the MCC trees, sharing some parameters that can be simultaneously estimated. This is where we take the timed phylogeny and use it to try to infer who-infected-whom. (Alternatively, you can use <code>run_TransPhylo</code> to run TransPhylo on a subsample of BEAST2 trees, without parameter sharing.)</li>
<li>
<strong>Running a regression on the TransPhylo results</strong>: <code>regression</code> runs linear or logistic regression using the TransPhylo results. This is where we draw inference about the relationship between some covariates and individual transmission probabilities.</li>
</ol>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Install the <code>beast2tpPipeline</code> package from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"jessalynnsebastian/beast2tpPipeline"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>There are some dependencies that should be automatically downloaded. Then, load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jessalynnsebastian.github.io/beast2tpPipeline/">beast2tpPipeline</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing beast2tpPipeline</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jessalyn Sebastian <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessalyn Sebastian.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
