<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create BEAST2 XML Files from a Template — create_xml_files • beast2tpPipeline</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create BEAST2 XML Files from a Template — create_xml_files"><meta name="description" content="Use a template XML file to create new XML files for BEAST2 input.
This function can replace sequences, sampling dates, mcmc iterations,
frequency of storing trees, and, for a uniform clock rate, the minimum
and maximum clock rates.
This was designed just to be used with the provided template file, though
it could theoretically be used with others. The template file was written
for tuberculosis data. It contains an adjustment for ascertainment bias due
to the SNPs, a Gamma site model, and HKY substitution model, strict clock,
Coalescent constant population model, uniform clock rate prior (min/max editable),
and lognormal(1, 1.25) freqParameter, kappa, and popsize priors.
The template xml contains placeholders:
SNP_FILE_NAME_HERE for the name of the SNP file
ALIGNMENT_INFORMATION_HERE for the sequence information
DATE_INFORMATION_HERE for the sampling dates
CLOCKRATE_MINIMUM_HERE for the minimum uniform clock rate
CLOCKRATE_MAXIMUM_HERE for the maximum uniform clock rate
CLOCKRATE_INITIAL_HERE for the initial clock rate
MCMC_ITERATIONS_HERE for the number of MCMC iterations
STORE_EVERY_HERE for the frequency of storing trees

"><meta property="og:description" content="Use a template XML file to create new XML files for BEAST2 input.
This function can replace sequences, sampling dates, mcmc iterations,
frequency of storing trees, and, for a uniform clock rate, the minimum
and maximum clock rates.
This was designed just to be used with the provided template file, though
it could theoretically be used with others. The template file was written
for tuberculosis data. It contains an adjustment for ascertainment bias due
to the SNPs, a Gamma site model, and HKY substitution model, strict clock,
Coalescent constant population model, uniform clock rate prior (min/max editable),
and lognormal(1, 1.25) freqParameter, kappa, and popsize priors.
The template xml contains placeholders:
SNP_FILE_NAME_HERE for the name of the SNP file
ALIGNMENT_INFORMATION_HERE for the sequence information
DATE_INFORMATION_HERE for the sampling dates
CLOCKRATE_MINIMUM_HERE for the minimum uniform clock rate
CLOCKRATE_MAXIMUM_HERE for the maximum uniform clock rate
CLOCKRATE_INITIAL_HERE for the initial clock rate
MCMC_ITERATIONS_HERE for the number of MCMC iterations
STORE_EVERY_HERE for the frequency of storing trees

"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">beast2tpPipeline</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Home"><span class="fa fa-home"></span></a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/mcc_tree_pipeline.html">Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees</a></li>
    <li><a class="dropdown-item" href="../articles/tree_sample_pipeline.html">Using the BEAST2 - TransPhylo Pipeline with a Sample of BEAST2 Iterations (Posterior Trees)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jessalynnsebastian/beast2tpPipeline" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create BEAST2 XML Files from a Template</h1>

      <div class="d-none name"><code>create_xml_files.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Use a template XML file to create new XML files for BEAST2 input.
This function can replace sequences, sampling dates, mcmc iterations,
frequency of storing trees, and, for a uniform clock rate, the minimum
and maximum clock rates.</p>
<p>This was designed just to be used with the provided template file, though
it could theoretically be used with others. The template file was written
for tuberculosis data. It contains an adjustment for ascertainment bias due
to the SNPs, a Gamma site model, and HKY substitution model, strict clock,
Coalescent constant population model, uniform clock rate prior (min/max editable),
and lognormal(1, 1.25) freqParameter, kappa, and popsize priors.</p>
<p>The template xml contains placeholders:</p><ul><li><p>SNP_FILE_NAME_HERE for the name of the SNP file</p></li>
<li><p>ALIGNMENT_INFORMATION_HERE for the sequence information</p></li>
<li><p>DATE_INFORMATION_HERE for the sampling dates</p></li>
<li><p>CLOCKRATE_MINIMUM_HERE for the minimum uniform clock rate</p></li>
<li><p>CLOCKRATE_MAXIMUM_HERE for the maximum uniform clock rate</p></li>
<li><p>CLOCKRATE_INITIAL_HERE for the initial clock rate</p></li>
<li><p>MCMC_ITERATIONS_HERE for the number of MCMC iterations</p></li>
<li><p>STORE_EVERY_HERE for the frequency of storing trees</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_xml_files</span><span class="op">(</span></span>
<span>  path_to_template <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"xml_template/0_xml_template.xml"</span>, package <span class="op">=</span></span>
<span>    <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>  <span class="va">seqs_list</span>,</span>
<span>  <span class="va">cluster_assignments</span>,</span>
<span>  mcmc_iterations <span class="op">=</span> <span class="fl">2e+07</span>,</span>
<span>  store_every <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  min_clockrate <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">8</span><span class="op">)</span>,</span>
<span>  max_clockrate <span class="op">=</span> <span class="fl">5</span> <span class="op">*</span> <span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>  init_clockrate <span class="op">=</span> <span class="fl">10</span> <span class="op">*</span> <span class="va">min_clockrate</span>,</span>
<span>  whole_genome_length <span class="op">=</span> <span class="fl">4.2</span> <span class="op">*</span> <span class="fl">10</span><span class="op">^</span><span class="fl">6</span>,</span>
<span>  <span class="va">out_dir</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-path-to-template">path_to_template<a class="anchor" aria-label="anchor" href="#arg-path-to-template"></a></dt>
<dd><p>The path to the template XML file.</p></dd>


<dt id="arg-seqs-list">seqs_list<a class="anchor" aria-label="anchor" href="#arg-seqs-list"></a></dt>
<dd><p>A list of matrices of DNA sequences (or SNPs), where
rownames correspond to sample IDs.</p></dd>


<dt id="arg-cluster-assignments">cluster_assignments<a class="anchor" aria-label="anchor" href="#arg-cluster-assignments"></a></dt>
<dd><p>A data frame of cluster assignments, including
<code>sample_id</code>, <code>cluster_name</code>, and <code>collectdt</code>, as output by
<code>assign_snp_clusters</code>. All sample IDs in the <code>seqs</code> matrices must be
present in this data frame.</p></dd>


<dt id="arg-mcmc-iterations">mcmc_iterations<a class="anchor" aria-label="anchor" href="#arg-mcmc-iterations"></a></dt>
<dd><p>The number of MCMC iterations to run.</p></dd>


<dt id="arg-store-every">store_every<a class="anchor" aria-label="anchor" href="#arg-store-every"></a></dt>
<dd><p>The frequency of storing trees.</p></dd>


<dt id="arg-min-clockrate">min_clockrate<a class="anchor" aria-label="anchor" href="#arg-min-clockrate"></a></dt>
<dd><p>The minimum clock rate for the uniform clock
rate prior. Defaults to tuberculosis genome clock rate. DO NOT adjust
for SNPs here.</p></dd>


<dt id="arg-max-clockrate">max_clockrate<a class="anchor" aria-label="anchor" href="#arg-max-clockrate"></a></dt>
<dd><p>The maximum clock rate for the uniform clock
rate prior. Defaults to tuberculosis genome clock rate.</p></dd>


<dt id="arg-init-clockrate">init_clockrate<a class="anchor" aria-label="anchor" href="#arg-init-clockrate"></a></dt>
<dd><p>The initial clock rate for the clock rate.</p></dd>


<dt id="arg-whole-genome-length">whole_genome_length<a class="anchor" aria-label="anchor" href="#arg-whole-genome-length"></a></dt>
<dd><p>The length of the whole genome. Used
to adjust the clock rate for SNPs. Defaults to TB length.</p></dd>


<dt id="arg-out-dir">out_dir<a class="anchor" aria-label="anchor" href="#arg-out-dir"></a></dt>
<dd><p>The directory to write the output XML file. Files
are automatically named after their FASTA names.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessalyn Sebastian.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

