<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees â€¢ beast2tpPipeline</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">beast2tpPipeline</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Home"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/mcc_tree_pipeline.html">Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jessalynnsebastian/beast2tpPipeline" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees</h1>
            
      

      <div class="d-none name"><code>mcc_tree_pipeline.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette will walk you through the process of using the
<code>beast2tpPipeline</code> package to create SNP clusters from a set
of fasta files, create xml files for each cluster, run BEAST2, run
TransPhylo on the BEAST2 results, and then do some kind of regression on
the probabilities assigned by TransPhylo.</p>
<p>After running BEAST2, we have a set of posterior trees for each
cluster. With this package we can do two things:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Create a <a href="https://en.wikipedia.org/wiki/Maximum_clade_credibility_tree#:~:text=A%20maximum%20clade%20credibility%20tree,of%20the%20sampled%20posterior%20trees.0" class="external-link">maximum
clade credibility (MCC) tree</a> from the posterior trees, and run
TransPhylo multitree on the MCC trees, sharing some parameters that can
be simultaneously estimated</strong>. The MCC tree is a single tree that
attempts to summarize the posterior trees, like a mean or median when
posterior samples are values from the real line. With this method we
take one tree that summarizes the posterior trees for each cluster, and
run TransPhylo on all of them as if they are the true phylogenetic
trees.</li>
<li>
<strong>Subsample some number of trees from the BEAST2 posterior
tree samples for each cluster, and run TransPhylo on a sample of trees
for each cluster, rather than one summary tree</strong>. This has the
advantage of accounting for some of the phylogenetic uncertainty in the
BEAST2 trees, rather than assuming that one tree is the truth. However,
it is computationally more expensive: instead of one TransPhylo run on
all the clusters, we have to run TransPhylo on each cluster separately,
each time with a decently-sized set of trees. This should be run on a
computing cluster. Additionally, via this method, parameter sharing is
not possible.</li>
</ol>
<p>In this vignette, we will use this package to do (1).</p>
<p>First, load the package with <code>library</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jessalynnsebastian.github.io/beast2tpPipeline/" class="external-link">beast2tpPipeline</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-for-this-example">Data for this Example<a class="anchor" aria-label="anchor" href="#data-for-this-example"></a>
</h2>
<p>This package comes with data collected from the Kopanyo study, in
which tuberculosis (TB) samples were collected in two regions of
Botswana, a country with high prevalence of both TB and human
immodeficiency virus (HIV). Read more about the study <a href="https://bmjopen.bmj.com/content/6/5/e010046" class="external-link">here</a>.</p>
<p>These data contain 4 fasta files, one for each lineage of TB. Each
fasta file contains aligned TB sequences from study participants. There
is also metadata available for these samples, including things like HIV
status, age, gender, sample collection date, and a handful of other
variables.</p>
<p>Suppose we are interested in using these TB sequences to draw
inference on whether having HIV makes an individual more likely to
transmit TB to someone else.</p>
</div>
<div class="section level2">
<h2 id="creating-single-nucleotide-polymorphism-snp-clusters">Creating Single Nucleotide Polymorphism (SNP) Clusters<a class="anchor" aria-label="anchor" href="#creating-single-nucleotide-polymorphism-snp-clusters"></a>
</h2>
<p>First, read the fasta files and the metadata into <code>R</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fasta_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"kopanyo"</span>, package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                          pattern <span class="op">=</span> <span class="st">".fasta"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tb_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fasta_files</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.dna</a></span>, format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>                       as.character <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"kopanyo"</span>, package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                       pattern <span class="op">=</span> <span class="st">".csv"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<p>We can take a peek at the data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   SampleID Lineage genderf1 agenew hivfinal_new gMixture  collectdt smokef1</span></span>
<span><span class="co">## 1 BTB-1139       1        1     52            1        0 2014-07-30       1</span></span>
<span><span class="co">## 2 BTB-1146       1        1     43            1        1 2014-08-05       1</span></span>
<span><span class="co">## 3 BTB-1705       1        1     70            2        1 2015-06-10       1</span></span>
<span><span class="co">## 4  BTB-175       1        2     19            1        1 2013-01-16       2</span></span>
<span><span class="co">## 5 BTB-1774       1        1     38            1        1 2015-07-30       2</span></span>
<span><span class="co">## 6 BTB-1894       1        1     15           NA        1 2015-10-16       2</span></span>
<span><span class="co">##   tb_everf3 alcohol_excess Lineage_detailed</span></span>
<span><span class="co">## 1         2              2            1.1.2</span></span>
<span><span class="co">## 2         2              1            1.1.2</span></span>
<span><span class="co">## 3         1              2            1.1.2</span></span>
<span><span class="co">## 4         1              2            1.1.2</span></span>
<span><span class="co">## 5         1              2            1.1.2</span></span>
<span><span class="co">## 6         1              2            1.1.2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1426   11</span></span></code></pre>
<p>Notice that we have 1426 sequences and corresponding metadata. It
would be very computationally intensive to run BEAST2 on all of these
sequences simultaneously. Instead, we will create clusters of similar
sequences and run BEAST2 on each cluster separately. To create clusters,
we will use SNP thresholding.</p>
<p>SNPs are single nucleotide polymorphisms, which are single base pair
differences between sequences. We can use these differences to group
sequences into clusters that are more likely to be closely related to
each other. We will use the <code>transcluster</code> package to do
this. <code>transcluster</code> is a package developed alongside the
paper <a href="https://academic.oup.com/mbe/article/36/3/587/5300248" class="external-link">Beyond the
SNP Threshold: Identifying Outbreak Clusters Using Inferred
Transmissions</a>. It was built to handle both plain SNP thresholding to
create clusters, and also clustering based on likely transmissions. Here
we will use the former, as it is simpler and more commonly used, but the
latter is still available in the function.</p>
<p>To assign sequences to their clusters based on SNPs, we will use
<code>assign_snp_clusters</code>, which is a wrapper around some
<code>transcluster</code> functions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get collection dates by lineage</span></span>
<span><span class="va">collectdts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">Lineage</span><span class="op">)</span></span>
<span><span class="co"># Create named vectors of dates, as required by the function</span></span>
<span><span class="va">collectdts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">collectdts</span>, <span class="kw">function</span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">dates</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">collectdt</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dates</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">SampleID</span></span>
<span>  <span class="va">dates</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># TODO: REMOVE THIS WHEN TESTING IS DONE.</span></span>
<span><span class="va">tb_sequences</span> <span class="op">&lt;-</span> <span class="va">tb_sequences</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">collectdts</span> <span class="op">&lt;-</span> <span class="va">collectdts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Apply the function to each lineage of sequences</span></span>
<span><span class="va">cluster_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">assign_snp_clusters</span>,</span>
<span>                              seqs <span class="op">=</span> <span class="va">tb_sequences</span>,</span>
<span>                              collectdts <span class="op">=</span> <span class="va">collectdts</span>,</span>
<span>                              threshold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                              SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating SNP-based clusters</span></span>
<span><span class="co">## Creating SNP-based clusters</span></span>
<span><span class="co">## Creating SNP-based clusters</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rename clusters to include lineage</span></span>
<span><span class="va">cluster_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">cluster_assignments</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cluster_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"lineage"</span>, <span class="va">i</span>, <span class="st">"_"</span>, <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cluster_name</span><span class="op">)</span></span>
<span>  <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the cluster assignments</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cluster_assignments</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1  BTB-1023 lineage1_cluster1 2014-05-13</span></span>
<span><span class="co">## 2  BTB-1058 lineage1_cluster2 2014-06-04</span></span>
<span><span class="co">## 3  BTB-1088 lineage1_cluster3 2014-06-25</span></span>
<span><span class="co">## 4    BTB-10 lineage1_cluster4 2012-09-10</span></span>
<span><span class="co">## 5  BTB-1133 lineage1_cluster5 2014-07-25</span></span>
<span><span class="co">## 6  BTB-1139 lineage1_cluster6 2014-07-30</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1  BTB-1000 lineage2_cluster1 2014-04-28</span></span>
<span><span class="co">## 2  BTB-1014 lineage2_cluster2 2014-05-07</span></span>
<span><span class="co">## 3  BTB-1025 lineage2_cluster3 2014-05-13</span></span>
<span><span class="co">## 4  BTB-1115 lineage2_cluster4 2014-07-16</span></span>
<span><span class="co">## 5  BTB-1123 lineage2_cluster5 2014-07-16</span></span>
<span><span class="co">## 6  BTB-1160 lineage2_cluster6 2014-08-12</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1   BTB-119 lineage3_cluster1 2012-12-03</span></span>
<span><span class="co">## 2  BTB-1259 lineage3_cluster2 2014-09-25</span></span>
<span><span class="co">## 3  BTB-1344 lineage3_cluster3 2014-11-11</span></span>
<span><span class="co">## 4  BTB-1410 lineage3_cluster1 2014-12-10</span></span>
<span><span class="co">## 5   BTB-155 lineage3_cluster1 2013-01-07</span></span>
<span><span class="co">## 6  BTB-1698 lineage3_cluster1 2015-06-08</span></span></code></pre>
<p>Computing the distance matrix for large clusters can take time, so if
you already have a distance matrix saved, you can pass it to
<code>assign_snp_clusters</code> with the <code>dist_matrix</code>
argument.</p>
<p>Next, we will create the actual BEAST2 clusters using the
<code>create_BEAST2_clusters</code> function. We will take the
sequences, split them into separate objects, keep only the SNPs (sites
that are the same across all sequences will not matter to BEAST2), and
write them to fasta files to be used by BEAUti or by
<code>create_cluster_xml</code>.</p>
<p>We will discard clusters with fewer than 4 sequences, and clusters
with fewer than 8 SNPs. We also add constant sites to the sequences, to
deal with ascertainment bias in BEAST2 since we are only keeping the
SNPs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For each lineage, use the cluster assignments to put the sequences</span></span>
<span><span class="co"># into clusters</span></span>
<span><span class="va">snp_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">tb_sequences</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">lineage_index</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Note that the TB sequences are in format matrix</span></span>
<span>  <span class="fu"><a href="../reference/create_BEAST2_clusters.html">create_BEAST2_clusters</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">tb_sequences</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                         <span class="co"># and snp_clusters is a named vector of cluster assignments,</span></span>
<span>                         <span class="co"># where the names are the sample ids</span></span>
<span>                         snp_clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cluster_name</span>,</span>
<span>                                                 <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span>,</span>
<span>                         min_cluster_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                         min_varsites <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                         snps_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         constant_sites <span class="op">=</span> <span class="st">"acgt"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># Now we have a list of 4 lineages, each containing a list of clusters</span></span>
<span><span class="co"># (data frames) within that lineage</span></span>
<span><span class="co"># Now that we won't be working with the fasta files themselves anymore,</span></span>
<span><span class="co"># let's turn this list of lists into a list, and let's turn the list of</span></span>
<span><span class="co"># cluster assignments into a data frame</span></span>
<span><span class="va">snp_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">snp_matrices</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">cluster_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">cluster_assignments</span><span class="op">)</span></span></code></pre></div>
<p>For each cluster, we will create an xml file for BEAST2. In general,
BEAUti is used to create xml files for BEAST2, but here we are using a
model that should be mostly the same across all clusters. For the TB
data, this package comes with a template xml and the function
<code>create_cluster_xml</code> that will fill in the necessary
information for each cluster. The model associated with this xml
template:</p>
<ul>
<li>Strict clock model</li>
<li>HKY substitution model</li>
<li>Coalescent constant population tree prior</li>
<li>Uniform clock rate prior, user-supplied bounds</li>
<li>(Default) Log-normal(1, 1.25) (w/ upper bound of 1) frequency
parameter</li>
<li>(Default) Log-normal(1, 1.25) transition-transversion parameter for
HKY (kappa)</li>
<li>(Default) Log-normal(1, 1.25) coalescent population size
parameter</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For each cluster, create an xml file for BEAST2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">snp_matrices</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cluster_name</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/create_cluster_xml.html">create_cluster_xml</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">snp_matrices</span><span class="op">[[</span><span class="va">cluster_name</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                     cluster_name <span class="op">=</span> <span class="va">cluster_name</span>,</span>
<span>                     sampling_dates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">cluster_assignments</span><span class="op">$</span><span class="va">collectdt</span><span class="op">[</span><span class="va">cluster_assignments</span><span class="op">$</span><span class="va">cluster_name</span> <span class="op">==</span> <span class="va">cluster_name</span><span class="op">]</span>,</span>
<span>                                               <span class="va">cluster_assignments</span><span class="op">$</span><span class="va">sample_id</span><span class="op">[</span><span class="va">cluster_assignments</span><span class="op">$</span><span class="va">cluster_name</span> <span class="op">==</span> <span class="va">cluster_name</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                     out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                           package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># TODO: make sure parallelizing works on windows</span></span>
<span><span class="va">cores</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="va">input_xml</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                    package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                        pattern <span class="op">=</span> <span class="st">"\\.xml"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="va">input_xml</span>, <span class="va">run_beast2</span>, mc.cores <span class="op">=</span> <span class="va">cores</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beast_logs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                     package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                         pattern <span class="op">=</span> <span class="st">"\\.log"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">beast_logs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">log</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/ess_checks.html">ess_checks</a></span><span class="op">(</span><span class="st">"BEAST2"</span>, <span class="va">log</span>, min_ess <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trees_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                      package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                          pattern <span class="op">=</span> <span class="st">".trees"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">out_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>, <span class="st">"mcctree"</span>,</span>
<span>                       package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">trees_files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/get_mcctree.html">get_mcctree</a></span><span class="op">(</span>input_treesfile <span class="op">=</span> <span class="va">trees_files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, output_dir <span class="op">=</span> <span class="va">out_dir</span><span class="op">)</span></span>
<span><span class="op">}</span>, mc.cores <span class="op">=</span> <span class="va">cores</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read the trees and manipulate the names to make sure they match</span></span>
<span><span class="co"># the names in cluster_assignments</span></span>
<span><span class="va">mcc_trees_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>,</span>
<span>                                          <span class="st">"BEAST2"</span>, <span class="st">"mcctree"</span>,</span>
<span>                                          package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                              pattern <span class="op">=</span> <span class="st">".tree"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mcc_trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">mcc_trees_files</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">read.nexus</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mcc_trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".nexus"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">mcc_trees_files</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mcc_trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*-"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mcc_trees</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Run TransPhylo on the MCC trees</span></span>
<span><span class="va">prob_source</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_TransPhylo.html">run_TransPhylo</a></span><span class="op">(</span><span class="va">mcc_trees</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"mcctrees"</span>,</span>
<span>                              cluster_dict <span class="op">=</span> <span class="va">cluster_assignments</span>,</span>
<span>                              out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>,</span>
<span>                                                    <span class="st">"TransPhylo"</span>,</span>
<span>                                                    package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                              w.shape <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                              w.scale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                              prior_pi_a <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                              prior_pi_b <span class="op">=</span> <span class="fl">19</span>,</span>
<span>                              share <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"neg"</span>, <span class="st">"off.r"</span><span class="op">)</span>,</span>
<span>                              startNeg <span class="op">=</span> <span class="fl">1.48</span>,</span>
<span>                              mcmcIterations <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prob_source</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"TransPhylo"</span>,</span>
<span>                                   <span class="st">"tp_res_prob_source.rds"</span>,</span>
<span>                                   package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Get dataframe with covariates for regression</span></span>
<span><span class="va">cleaned_data</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"hivfinal_new"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cleaned_data</span> <span class="op">&lt;-</span> <span class="va">cleaned_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">cleaned_data</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">cleaned_data</span><span class="op">$</span><span class="va">hivfinal_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">cleaned_data</span><span class="op">$</span><span class="va">hivfinal_new</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Run a linear model with the probabilities from TransPhylo</span></span>
<span><span class="va">mdl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regression.html">regression</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"linear"</span>,</span>
<span>                  cleaned_data <span class="op">=</span> <span class="va">cleaned_data</span>,</span>
<span>                  prob_source <span class="op">=</span> <span class="va">prob_source</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mdl</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessalyn Sebastian.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
