<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees • beast2tpPipeline</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">beast2tpPipeline</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="Home"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/mcc_tree_pipeline.html">Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees</a></li>
    <li><a class="dropdown-item" href="../articles/tree_sample_pipeline.html">Using the BEAST2 - TransPhylo Pipeline with a Sample of BEAST2 Iterations (Posterior Trees)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jessalynnsebastian/beast2tpPipeline" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Using the BEAST2 - TransPhylo Pipeline with Maximum Clade Credibility Trees</h1>
            
      

      <div class="d-none name"><code>mcc_tree_pipeline.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette will walk you through the process of using the
<code>beast2tpPipeline</code> package to create SNP clusters from a set
of fasta files, create xml files for each cluster, run BEAST2, run
TransPhylo on the BEAST2 results, and then do some kind of regression on
the probabilities assigned by TransPhylo.</p>
<p>After running BEAST2, we have a set of posterior trees for each
cluster. With this package we can do two things:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Create a <a href="https://en.wikipedia.org/wiki/Maximum_clade_credibility_tree#:~:text=A%20maximum%20clade%20credibility%20tree,of%20the%20sampled%20posterior%20trees.0" class="external-link">maximum
clade credibility (MCC) tree</a> from the posterior trees, and run
TransPhylo multitree on the MCC trees, sharing some parameters that can
be simultaneously estimated</strong>. The MCC tree is a single tree that
attempts to summarize the posterior trees, like a mean or median when
posterior samples are values from the real line. With this method we
take one tree that summarizes the posterior trees for each cluster, and
run TransPhylo on all of them as if they are the true phylogenetic
trees.</li>
<li>
<strong>Subsample some number of trees from the BEAST2 posterior
tree samples for each cluster, and run TransPhylo on a sample of trees
for each cluster, rather than one summary tree</strong>. This has the
advantage of accounting for some of the phylogenetic uncertainty in the
BEAST2 trees, rather than assuming that one tree is the truth. However,
it is computationally more expensive: instead of one TransPhylo run on
all the clusters, we have to run TransPhylo on each cluster separately,
each time with a decently-sized set of trees. This should be run on a
computing cluster. Additionally, via this method, parameter sharing is
not possible.</li>
</ol>
<p>In this vignette, we will use this package to do (1). For the
vignette on (2), see Using the BEAST2 - TransPhylo Pipeline with a
Sample of BEAST2 Iterations (Posterior Trees).</p>
<p>First, load the package with <code>library</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jessalynnsebastian.github.io/beast2tpPipeline/">beast2tpPipeline</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-for-this-example">Data for this Example<a class="anchor" aria-label="anchor" href="#data-for-this-example"></a>
</h2>
<p>This package comes with data collected from the Kopanyo study, in
which tuberculosis (TB) samples were collected in two regions of
Botswana, a country with high prevalence of both TB and human
immodeficiency virus (HIV). Read more about the study <a href="https://bmjopen.bmj.com/content/6/5/e010046" class="external-link">here</a>.</p>
<p>These data contain 4 fasta files, one for each lineage of TB. Each
fasta file contains aligned TB sequences from study participants. There
is also metadata available for these samples, including things like HIV
status, age, gender, sample collection date, and a handful of other
variables.</p>
<p>Suppose we are interested in using these TB sequences to draw
inference on whether having HIV makes an individual more likely to
transmit TB to someone else.</p>
</div>
<div class="section level2">
<h2 id="creating-single-nucleotide-polymorphism-snp-clusters">Creating Single Nucleotide Polymorphism (SNP) Clusters<a class="anchor" aria-label="anchor" href="#creating-single-nucleotide-polymorphism-snp-clusters"></a>
</h2>
<p>First, read the fasta files and the metadata into <code>R</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fasta_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"kopanyo"</span>, package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                          pattern <span class="op">=</span> <span class="st">".fasta"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tb_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fasta_files</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.dna</a></span>, format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>                       as.character <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"kopanyo"</span>, package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                       pattern <span class="op">=</span> <span class="st">".csv"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<p>We can take a peek at the data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   SampleID Lineage genderf1 agenew hivfinal_new gMixture  collectdt smokef1</span></span>
<span><span class="co">## 1 BTB-1139       1        1     52            1        0 2014-07-30       1</span></span>
<span><span class="co">## 2 BTB-1146       1        1     43            1        1 2014-08-05       1</span></span>
<span><span class="co">## 3 BTB-1705       1        1     70            2        1 2015-06-10       1</span></span>
<span><span class="co">## 4  BTB-175       1        2     19            1        1 2013-01-16       2</span></span>
<span><span class="co">## 5 BTB-1774       1        1     38            1        1 2015-07-30       2</span></span>
<span><span class="co">## 6 BTB-1894       1        1     15           NA        1 2015-10-16       2</span></span>
<span><span class="co">##   tb_everf3 alcohol_excess Lineage_detailed</span></span>
<span><span class="co">## 1         2              2            1.1.2</span></span>
<span><span class="co">## 2         2              1            1.1.2</span></span>
<span><span class="co">## 3         1              2            1.1.2</span></span>
<span><span class="co">## 4         1              2            1.1.2</span></span>
<span><span class="co">## 5         1              2            1.1.2</span></span>
<span><span class="co">## 6         1              2            1.1.2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1426   11</span></span></code></pre>
<p>Notice that we have 1426 sequences and corresponding metadata. It
would be very computationally intensive to run BEAST2 on all of these
sequences simultaneously. Instead, we will create clusters of similar
sequences and run BEAST2 on each cluster separately. To create clusters,
we will use SNP thresholding.</p>
<p>SNPs are single nucleotide polymorphisms, which are single base pair
differences between sequences. We can use these differences to group
sequences into clusters that are more likely to be closely related to
each other. We will use the <code>transcluster</code> package to do
this. <code>transcluster</code> is a package developed alongside the
paper <a href="https://academic.oup.com/mbe/article/36/3/587/5300248" class="external-link">Beyond the
SNP Threshold: Identifying Outbreak Clusters Using Inferred
Transmissions</a>. It was built to handle both plain SNP thresholding to
create clusters, and also clustering based on likely transmissions. Here
we will use the former, as it is simpler and more commonly used, but the
latter is still available in the function.</p>
<p>To assign sequences to their clusters based on SNPs, we will use
<code>assign_snp_clusters</code>, which is a wrapper around some
<code>transcluster</code> functions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get collection dates by lineage</span></span>
<span><span class="va">collectdts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">Lineage</span><span class="op">)</span></span>
<span><span class="co"># Create named vectors of dates, as required by the function</span></span>
<span><span class="va">collectdts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">collectdts</span>, <span class="kw">function</span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">dates</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">collectdt</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dates</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">SampleID</span></span>
<span>  <span class="va">dates</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Apply the function to each lineage of sequences</span></span>
<span><span class="va">cluster_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">assign_snp_clusters</span>,</span>
<span>                              <span class="co"># `seqs` is a matrix of sequences</span></span>
<span>                              seqs <span class="op">=</span> <span class="va">tb_sequences</span>,</span>
<span>                              <span class="co"># `collectdts` is a named list of vectors of collection dates</span></span>
<span>                              <span class="co"># where the elements are collection dates and the names are sample IDs  </span></span>
<span>                              collectdts <span class="op">=</span> <span class="va">collectdts</span>,</span>
<span>                              threshold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                              SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating SNP-based clusters</span></span>
<span><span class="co">## Creating SNP-based clusters</span></span>
<span><span class="co">## Creating SNP-based clusters</span></span>
<span><span class="co">## Creating SNP-based clusters</span></span></code></pre>
<p>Computing the distance matrix for large clusters can take time, so if
you already have a distance matrix saved, you can pass it to
<code>assign_snp_clusters</code> with the <code>dist_matrix</code>
argument.</p>
<p>Now we have:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rename clusters to include lineage</span></span>
<span><span class="va">cluster_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">cluster_assignments</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cluster_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"lineage"</span>, <span class="va">i</span>, <span class="st">"_"</span>, <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cluster_name</span><span class="op">)</span></span>
<span>  <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the cluster assignments</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cluster_assignments</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1  BTB-1023 lineage1_cluster1 2014-05-13</span></span>
<span><span class="co">## 2  BTB-1058 lineage1_cluster2 2014-06-04</span></span>
<span><span class="co">## 3  BTB-1088 lineage1_cluster3 2014-06-25</span></span>
<span><span class="co">## 4    BTB-10 lineage1_cluster4 2012-09-10</span></span>
<span><span class="co">## 5  BTB-1133 lineage1_cluster5 2014-07-25</span></span>
<span><span class="co">## 6  BTB-1139 lineage1_cluster6 2014-07-30</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1  BTB-1000 lineage2_cluster1 2014-04-28</span></span>
<span><span class="co">## 2  BTB-1014 lineage2_cluster2 2014-05-07</span></span>
<span><span class="co">## 3  BTB-1025 lineage2_cluster3 2014-05-13</span></span>
<span><span class="co">## 4  BTB-1115 lineage2_cluster4 2014-07-16</span></span>
<span><span class="co">## 5  BTB-1123 lineage2_cluster5 2014-07-16</span></span>
<span><span class="co">## 6  BTB-1160 lineage2_cluster6 2014-08-12</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1   BTB-119 lineage3_cluster1 2012-12-03</span></span>
<span><span class="co">## 2  BTB-1259 lineage3_cluster2 2014-09-25</span></span>
<span><span class="co">## 3  BTB-1344 lineage3_cluster3 2014-11-11</span></span>
<span><span class="co">## 4  BTB-1410 lineage3_cluster1 2014-12-10</span></span>
<span><span class="co">## 5   BTB-155 lineage3_cluster1 2013-01-07</span></span>
<span><span class="co">## 6  BTB-1698 lineage3_cluster1 2015-06-08</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">##   sample_id      cluster_name  collectdt</span></span>
<span><span class="co">## 1  BTB-1001 lineage4_cluster1 2014-04-28</span></span>
<span><span class="co">## 2  BTB-1003 lineage4_cluster2 2014-04-25</span></span>
<span><span class="co">## 3  BTB-1005 lineage4_cluster3 2014-04-30</span></span>
<span><span class="co">## 4  BTB-1007 lineage4_cluster4 2014-05-02</span></span>
<span><span class="co">## 5  BTB-1009 lineage4_cluster5 2014-05-05</span></span>
<span><span class="co">## 6  BTB-1010 lineage4_cluster6 2014-05-05</span></span></code></pre>
<p>Next, we will create the actual BEAST2 clusters containing the
genetic data using the <code>create_BEAST2_clusters</code> function. We
will take the sequences, split them into separate objects, and keep only
the SNPs (sites that are the same across all sequences will not matter
to BEAST2).</p>
<p>We will discard clusters with fewer than 4 sequences, and clusters
with fewer than 8 SNPs. We also add constant sites to the sequences, to
deal with ascertainment bias in BEAST2 due to the fact that we are only
keeping the SNPs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For each lineage, use the cluster assignments to put the sequences</span></span>
<span><span class="co"># into clusters</span></span>
<span><span class="va">snp_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">tb_sequences</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">lineage_index</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Pull the sequences and cluster assignments for this lineage</span></span>
<span>  <span class="va">tb_seq_lineage</span> <span class="op">&lt;-</span> <span class="va">tb_sequences</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">cluster_assignments_lineage</span> <span class="op">&lt;-</span> <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="../reference/create_BEAST2_clusters.html">create_BEAST2_clusters</a></span><span class="op">(</span>seqs <span class="op">=</span> <span class="va">tb_seq_lineage</span>,</span>
<span>                         cluster_assignments <span class="op">=</span> <span class="va">cluster_assignments_lineage</span>,</span>
<span>                         min_cluster_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                         min_varsites <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                         snps_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         constant_sites <span class="op">=</span> <span class="st">"acgt"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the clusters from the first lineage</span></span>
<span><span class="va">snp_matrices</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $lineage1_cluster6</span></span>
<span><span class="co">##          [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]</span></span>
<span><span class="co">## BTB-1139 "a"  "c"  "g"  "t"  "c"  "c"  "g"  "c"  "t"  "c"   "c"   "t"   "c"  </span></span>
<span><span class="co">## BTB-1146 "a"  "c"  "g"  "t"  "a"  "t"  "g"  "c"  "t"  "t"   "t"   "t"   "c"  </span></span>
<span><span class="co">## BTB-1705 "a"  "c"  "g"  "t"  "c"  "c"  "g"  "t"  "t"  "c"   "c"   "t"   "c"  </span></span>
<span><span class="co">## BTB-175  "a"  "c"  "g"  "t"  "c"  "c"  "g"  "c"  "t"  "c"   "c"   "c"   "t"  </span></span>
<span><span class="co">## BTB-1774 "a"  "c"  "g"  "t"  "c"  "c"  "c"  "c"  "t"  "c"   "c"   "c"   "t"  </span></span>
<span><span class="co">## BTB-262  "a"  "c"  "g"  "t"  "c"  "c"  "g"  "c"  "t"  "c"   "c"   "t"   "c"  </span></span>
<span><span class="co">## BTB-604  "a"  "c"  "g"  "t"  "c"  "c"  "g"  "c"  "c"  "c"   "c"   "t"   "c"  </span></span>
<span><span class="co">##          [,14] [,15] [,16] [,17] [,18]</span></span>
<span><span class="co">## BTB-1139 "g"   "g"   "c"   "c"   "g"  </span></span>
<span><span class="co">## BTB-1146 "g"   "g"   "c"   "c"   "a"  </span></span>
<span><span class="co">## BTB-1705 "g"   "g"   "c"   "t"   "g"  </span></span>
<span><span class="co">## BTB-175  "a"   "t"   "t"   "c"   "g"  </span></span>
<span><span class="co">## BTB-1774 "a"   "t"   "t"   "c"   "g"  </span></span>
<span><span class="co">## BTB-262  "g"   "g"   "c"   "c"   "g"  </span></span>
<span><span class="co">## BTB-604  "g"   "g"   "c"   "c"   "g"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $lineage1_cluster7</span></span>
<span><span class="co">##          [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]</span></span>
<span><span class="co">## BTB-1332 "a"  "c"  "g"  "t"  "c"  "c"  "c"  "g"  "a"  "t"   "t"   "c"   "a"  </span></span>
<span><span class="co">## BTB-1645 "a"  "c"  "g"  "t"  "t"  "a"  "c"  "t"  "a"  "t"   "t"   "c"   "g"  </span></span>
<span><span class="co">## BTB-584  "a"  "c"  "g"  "t"  "c"  "c"  "c"  "g"  "a"  "t"   "t"   "g"   "a"  </span></span>
<span><span class="co">## BTB-640  "a"  "c"  "g"  "t"  "c"  "c"  "c"  "g"  "g"  "c"   "t"   "c"   "g"  </span></span>
<span><span class="co">## BTB-665  "a"  "c"  "g"  "t"  "c"  "c"  "g"  "g"  "a"  "t"   "a"   "c"   "a"  </span></span>
<span><span class="co">##          [,14]</span></span>
<span><span class="co">## BTB-1332 "g"  </span></span>
<span><span class="co">## BTB-1645 "g"  </span></span>
<span><span class="co">## BTB-584  "g"  </span></span>
<span><span class="co">## BTB-640  "a"  </span></span>
<span><span class="co">## BTB-665  "g"</span></span></code></pre>
<p>The output of this function is a list of matrices, where each matrix
is a cluster of sequences. We will use these to write into the xml file
to run BEAST2. If you are not using the template xml file available with
this package, you should save these sequences as fasta files to be
loaded into BEAUti, which you can do by providing a
<code>fasta_dir</code> argument to the above function.</p>
<p>After running the above chunks, we have two objects:
<code>cluster_assignments</code>, a list of lineages each containing a
dataframe giving sample IDs, cluster names, and collection dates; and
<code>snp_matrices</code>, a list of lineages each containing a list of
matrices, where each matrix is a cluster of sequences.</p>
</div>
<div class="section level2">
<h2 id="creating-xml-files-for-beast2">Creating XML Files for BEAST2<a class="anchor" aria-label="anchor" href="#creating-xml-files-for-beast2"></a>
</h2>
<p>For each cluster, we will create an xml file for BEAST2. In general,
BEAUti is used to create xml files for BEAST2, but here we are using a
model that should be mostly the same across all clusters. For the TB
data, this package comes with a template xml and the function
<code>create_cluster_xml</code> that will fill in the necessary
information for each cluster. The model associated with this xml
template:</p>
<ul>
<li>Strict clock model</li>
<li>HKY substitution model</li>
<li>Coalescent constant population tree prior</li>
<li>Uniform clock rate prior, user-supplied bounds</li>
<li>Log-normal(1, 1.25) (w/ upper bound of 1) frequency parameter</li>
<li>Log-normal(1, 1.25) transition-transversion parameter for HKY
(kappa)</li>
<li>Log-normal(1, 1.25) coalescent population size parameter</li>
</ul>
<p>The function to create the xml files requires the sequences (SNPs,
for us) for the cluster, the name of the cluster, the sampling dates for
the sequences in the cluster, and a directory to output the xml files
to.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For each lineage, write the xml files for all clusters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">snp_matrices</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">lineage_index</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Get list of clusters for the lineage</span></span>
<span>  <span class="va">lineage_seqs</span> <span class="op">&lt;-</span> <span class="va">snp_matrices</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="co"># Get cluster assignments data frame for the lineage</span></span>
<span>  <span class="va">cluster_assignments_lineage</span> <span class="op">&lt;-</span> <span class="va">cluster_assignments</span><span class="op">[[</span><span class="va">lineage_index</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="../reference/create_xml_files.html">create_xml_files</a></span><span class="op">(</span>seqs_list <span class="op">=</span> <span class="va">lineage_seqs</span>,</span>
<span>                   cluster_assignments <span class="op">=</span> <span class="va">cluster_assignments_lineage</span>,</span>
<span>                   out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                         package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The function above does not return anything, but it writes xml files
to the directory specified in <code>out_dir</code>. Messages are printed
when each xml file is written.</p>
</div>
<div class="section level2">
<h2 id="running-beast2">Running BEAST2<a class="anchor" aria-label="anchor" href="#running-beast2"></a>
</h2>
<p>Next, we will run BEAST2 on the xml files we just created. We will
use the <code>run_beast2</code> function, which calls command-line
BEAST2 using <code>system</code>. You may need to change the
<code>beast2_path</code> argument depending on which version of BEAST2
you have installed and where it is installed to.</p>
<p>Below I have parallelized the code using the <code>parallel</code>
package. This is not necessary, but it can speed up the process if you
have multiple cores available. Note that this will not work on Windows,
but there are other ways to parallelize on Windows machines.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cores</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="va">input_xml</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                    package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                        pattern <span class="op">=</span> <span class="st">"\\.xml"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="va">input_xml</span>, <span class="va">run_beast2</span>, mc.cores <span class="op">=</span> <span class="va">cores</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>After running BEAST2, for each cluster, we will have a log file and a
tree file. The log file logs the MCMC sampling process, and the tree
file contains the posterior trees. First, as a quick check for the
mixing of the MCMC, we can make sure the effective sample size (ESS) is
above 200 for all parameters. We can do this using the
<code>ess_checks</code> function, which calls on the package
<code>tracerer</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beast_logs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                     package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                         pattern <span class="op">=</span> <span class="st">"\\.log"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">beast_logs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">log</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/ess_checks.html">ess_checks</a></span><span class="op">(</span><span class="st">"BEAST2"</span>, <span class="va">log</span>, min_ess <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span></code></pre>
<pre><code><span><span class="co">## lineage4_cluster183.log:</span></span>
<span><span class="co">## ESS is too low for 2 parameters</span></span></code></pre>
<pre><code><span><span class="co">##  ESS: 117    ESS: 162</span></span></code></pre>
<pre><code><span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span>
<span><span class="co">## All ESS are above minimum threshold</span></span></code></pre>
<p>If the ESS is too low, you may want to consider re-running BEAST2
with more iterations. Be sure also to look at traceplots of the MCMC
run, which is easiest to do with <code>Tracer</code>.</p>
</div>
<div class="section level2">
<h2 id="creating-maximum-clade-credibility-trees">Creating Maximum Clade Credibility Trees<a class="anchor" aria-label="anchor" href="#creating-maximum-clade-credibility-trees"></a>
</h2>
<p>Next, we will create a maximum clade credibility (MCC) tree from the
posterior trees. The MCC tree is a single tree that attempts to
summarize the posterior trees, like a mean or median when posterior
samples are values from the real line. We will use the
<code>get_mcctree</code> function for this. <code>get_mcctree</code>
calls command-line TreeAnnotator from the BEAST2 suite, so you may need
to change the <code>treeannotator_path</code> argument depending on
where TreeAnnotator is installed.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trees_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>,</span>
<span>                                      package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                          pattern <span class="op">=</span> <span class="st">".trees"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">out_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>, <span class="st">"BEAST2"</span>, <span class="st">"mcctree"</span>,</span>
<span>                       package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">trees_files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/get_mcctree.html">get_mcctree</a></span><span class="op">(</span>input_treesfile <span class="op">=</span> <span class="va">trees_files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, output_dir <span class="op">=</span> <span class="va">out_dir</span><span class="op">)</span></span>
<span><span class="op">}</span>, mc.cores <span class="op">=</span> <span class="va">cores</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This function outputs nothing, but MCC trees are saved to the
specified output directory.</p>
<p>The MCC trees look like this, for example:</p>
<p><img src="mcc_tree_pipeline_files/figure-html/mcc_tree_preview-1.png" width="700"></p>
<p>where there are associated error bars for each node and uncertainty
in the tree topology, but this will not be accounted for by
TransPhylo.</p>
</div>
<div class="section level2">
<h2 id="running-transphylo-on-the-mcc-trees">Running TransPhylo on the MCC Trees<a class="anchor" aria-label="anchor" href="#running-transphylo-on-the-mcc-trees"></a>
</h2>
<p>Now that we have the MCC trees, we can run TransPhylo on them.
TransPhylo is a package that tries to infer the “who infected whom”
tree, or transmission tree, taking the phylogenetic tree as data. In
reality, there is uncertainty in the phylogenetic tree, but TransPhylo
does not account for this.</p>
<p>We will use the <code>run_TransPhylo</code> function to run
TransPhylo on the MCC trees. We first read in the MCC trees we created
above, make sure their names match the names in the cluster assignments
data frame (BEAST2 does some slight renaming of its outputs
automatically), and then run TransPhylo. We can run TransPhylo on all of
the trees at once, sharing a couple of parameters that are common across
the trees and can be estimated simultaneously.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read the trees and manipulate the names to make sure they match</span></span>
<span><span class="co"># the names in cluster_assignments</span></span>
<span><span class="va">mcc_trees_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>,</span>
<span>                                          <span class="st">"BEAST2"</span>, <span class="st">"mcctree"</span>,</span>
<span>                                          package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                              pattern <span class="op">=</span> <span class="st">".tree"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mcc_trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">mcc_trees_files</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">read.nexus</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mcc_trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".nexus"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">mcc_trees_files</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mcc_trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*-"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mcc_trees</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Want to run trees from all lineages together, so paste the cluster</span></span>
<span><span class="co"># assignments list of dataframes into one dataframe</span></span>
<span><span class="va">all_cluster_assignments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">cluster_assignments</span><span class="op">)</span></span>
<span><span class="co"># Run TransPhylo on the MCC trees</span></span>
<span><span class="va">prob_source</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_TransPhylo.html">run_TransPhylo</a></span><span class="op">(</span><span class="va">mcc_trees</span>,</span>
<span>                              type <span class="op">=</span> <span class="st">"mcctrees"</span>,</span>
<span>                              cluster_dict <span class="op">=</span> <span class="va">all_cluster_assignments</span>,</span>
<span>                              out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>,</span>
<span>                                                    <span class="st">"TransPhylo"</span>,</span>
<span>                                                    package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                              w.shape <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                              w.scale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                              prior_pi_a <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                              prior_pi_b <span class="op">=</span> <span class="fl">19</span>,</span>
<span>                              share <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"neg"</span>, <span class="st">"off.r"</span><span class="op">)</span>,</span>
<span>                              startNeg <span class="op">=</span> <span class="fl">1.48</span>,</span>
<span>                              mcmcIterations <span class="op">=</span> <span class="fl">100000</span><span class="op">)</span></span></code></pre></div>
<p>This function outputs a data frame containing each sample ID and the
probability of that sample being an infection source, i.e., the
probability that they infected someone else with TB. It also saves the
TransPhylo results (the <code>resTransPhylo</code> object output by
<code><a href="https://rdrr.io/pkg/TransPhylo/man/infer_multittree_share_param.html" class="external-link">TransPhylo::infer_multittree_share_param</a></code>) as an
<code>.rds</code> file in the specified output directory.</p>
<p>We can also use the <code>ess_checks</code> function to check the
mixing of the TransPhylo run, but it is probably more useful to look at
the traceplots for estimated parameters. Below is an example of how to
do this, using a TransPhylo function with only the first cluster.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">transphylo_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">transphylo_res_file</span><span class="op">)</span></span>
<span><span class="fu">TransPhylo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TransPhylo/man/plotTraces.html" class="external-link">plotTraces</a></span><span class="op">(</span><span class="va">transphylo_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/example_traceplots.svg"><!-- --></p>
<p>TransPhylo also provides methods of summarizing the posterior trees,
and plotting the result. For example, we can get the “medoid” (extension
of median) tree and plot it. Here, the phylogenetic tree is plotted with
colors, where each color represents an individual and asterisks
represent transmission events. In TransPhylo this object is called a
<code>ctree</code>, or combined phylogenetic/transmission tree (“colored
tree”).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">medoid</span> <span class="op">&lt;-</span> <span class="fu">TransPhylo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TransPhylo/man/medTTree.html" class="external-link">medTTree</a></span><span class="op">(</span><span class="va">transphylo_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">TransPhylo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TransPhylo/man/plotCTree.html" class="external-link">plotCTree</a></span><span class="op">(</span><span class="va">medoid</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/medttree.svg"><!-- --></p>
<p>There is also a “consensus” tree, which is built by combining clades
occurring in at least some fraction of the posterior samples. This tree
will not necessarily be a tree that is in the posterior samples. It can
be plotted with <code>plot</code>. In TransPhylo this object is called a
<code>ttree</code>, or transmission tree.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cons</span> <span class="op">&lt;-</span> <span class="fu">TransPhylo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TransPhylo/man/consTTree.html" class="external-link">consTTree</a></span><span class="op">(</span><span class="va">transphylo_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cons</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/consttree.svg"><!-- --></p>
<p>We may also be interested in looking at the distribution of the
probabilities assigned by TransPhylo:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">prob_source</span><span class="op">$</span><span class="va">prob</span>, main <span class="op">=</span> <span class="st">"Distribution of TransPhylo Probabilities"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Probability of being an infector"</span>, col <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="mcc_tree_pipeline_files/figure-html/prob_source_histogram-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="regression-with-transphylo-results">Regression with TransPhylo Results<a class="anchor" aria-label="anchor" href="#regression-with-transphylo-results"></a>
</h2>
<p>Finally, we can do some kind of regression with the probabilities
assigned by TransPhylo. Here we will run a linear regression with the
probabilities as the response variable and HIV status as the predictor.
We will use the <code>regression</code> function to do so.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get dataframe with covariates for regression</span></span>
<span><span class="va">cleaned_data</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"hivfinal_new"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cleaned_data</span> <span class="op">&lt;-</span> <span class="va">cleaned_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">cleaned_data</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">cleaned_data</span><span class="op">$</span><span class="va">hivfinal_new</span> <span class="op">&lt;-</span> <span class="va">cleaned_data</span><span class="op">$</span><span class="va">hivfinal_new</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="co"># Run a linear model with the probabilities from TransPhylo</span></span>
<span><span class="va">mdl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regression.html">regression</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"linear"</span>,</span>
<span>                  cleaned_data <span class="op">=</span> <span class="va">cleaned_data</span>,</span>
<span>                  prob_source <span class="op">=</span> <span class="va">prob_source</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mdl</span><span class="op">)</span></span></code></pre></div>
<p>We can add more covariates by adding columns to the
<code>cleaned_data</code> data frame. The <code>regression</code>
function will automatically remove rows with missing values.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get dataframe with covariates for regression, add age and gender</span></span>
<span><span class="va">cleaned_data</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"hivfinal_new"</span>, <span class="st">"genderf1"</span>,</span>
<span>                             <span class="st">"tb_everf3"</span>, <span class="st">"smokef1"</span>, <span class="st">"alcohol_excess"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cleaned_data</span> <span class="op">&lt;-</span> <span class="va">cleaned_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">cleaned_data</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">cleaned_data</span><span class="op">[</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">cleaned_data</span><span class="op">[</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, <span class="fl">2</span>,</span>
<span>                             <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Run a linear model with the probabilities from TransPhylo</span></span>
<span><span class="va">mdl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regression.html">regression</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"linear"</span>,</span>
<span>                  cleaned_data <span class="op">=</span> <span class="va">cleaned_data</span>,</span>
<span>                  prob_source <span class="op">=</span> <span class="va">prob_source</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mdl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = prob_source ~ ., data = cleaned_data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residuals:</span></span>
<span><span class="co">##      Min       1Q   Median       3Q      Max </span></span>
<span><span class="co">## -0.22528 -0.17311 -0.15239  0.01174  0.83791 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##                 Estimate Std. Error t value Pr(&gt;|t|)  </span></span>
<span><span class="co">## (Intercept)     0.126011   0.057209   2.203   0.0286 *</span></span>
<span><span class="co">## hivfinal_new    0.039487   0.040426   0.977   0.3297  </span></span>
<span><span class="co">## genderf1       -0.011014   0.043591  -0.253   0.8007  </span></span>
<span><span class="co">## tb_everf3       0.007609   0.049439   0.154   0.8778  </span></span>
<span><span class="co">## smokef1         0.052170   0.051089   1.021   0.3082  </span></span>
<span><span class="co">## alcohol_excess -0.020719   0.049316  -0.420   0.6748  </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residual standard error: 0.3077 on 234 degrees of freedom</span></span>
<span><span class="co">## Multiple R-squared:  0.008643,   Adjusted R-squared:  -0.01254 </span></span>
<span><span class="co">## F-statistic: 0.408 on 5 and 234 DF,  p-value: 0.843</span></span></code></pre>
<p>Alternatively, a probability threshold can be set to mark individuals
as infectors or non-infectors, and a logistic regression may be run.
This is done by setting the <code>method</code> argument in
<code>regression</code> to <code>"logistic"</code> and specifying a
threshold.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run a logistic model with the probabilities from TransPhylo</span></span>
<span><span class="co"># Set the probability threshold to 0.6, where prob &gt;= 0.6 is</span></span>
<span><span class="co"># considered an infector</span></span>
<span><span class="va">mdl_logistic</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regression.html">regression</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"logistic"</span>,</span>
<span>                           cleaned_data <span class="op">=</span> <span class="va">cleaned_data</span>,</span>
<span>                           prob_source <span class="op">=</span> <span class="va">prob_source</span>,</span>
<span>                           prob_cutoff <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mdl_logistic</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## glm(formula = tp_source ~ ., family = "binomial", data = cleaned_data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##                Estimate Std. Error z value Pr(&gt;|z|)    </span></span>
<span><span class="co">## (Intercept)    -2.57277    0.61177  -4.205 2.61e-05 ***</span></span>
<span><span class="co">## hivfinal_new    0.61646    0.40656   1.516    0.129    </span></span>
<span><span class="co">## genderf1       -0.03172    0.42411  -0.075    0.940    </span></span>
<span><span class="co">## tb_everf3      -0.10814    0.50432  -0.214    0.830    </span></span>
<span><span class="co">## smokef1         1.03882    0.55876   1.859    0.063 .  </span></span>
<span><span class="co">## alcohol_excess -0.52863    0.47714  -1.108    0.268    </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Dispersion parameter for binomial family taken to be 1)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Null deviance: 184.70  on 239  degrees of freedom</span></span>
<span><span class="co">## Residual deviance: 178.82  on 234  degrees of freedom</span></span>
<span><span class="co">## AIC: 190.82</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Fisher Scoring iterations: 5</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get odds ratios (adjusted for other covariates)</span></span>
<span><span class="va">adj_odds_ratios</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>est <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">mdl_logistic</span><span class="op">$</span><span class="va">coefficients</span><span class="op">)</span>,</span>
<span>                              row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mdl_logistic</span><span class="op">$</span><span class="va">coefficients</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">adj_odds_ratios</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">adj_odds_ratios</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/confint.html" class="external-link">confint</a></span><span class="op">(</span><span class="va">mdl_logistic</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Waiting for profiling to be done...</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj_odds_ratios</span></span></code></pre></div>
<pre><code><span><span class="co">##                       est      2.5 %    97.5 %</span></span>
<span><span class="co">## (Intercept)    0.07632382 0.02069694 0.2330317</span></span>
<span><span class="co">## hivfinal_new   1.85236668 0.84517253 4.2139991</span></span>
<span><span class="co">## genderf1       0.96877762 0.41671834 2.2236658</span></span>
<span><span class="co">## tb_everf3      0.89750391 0.34946563 2.6086441</span></span>
<span><span class="co">## smokef1        2.82587567 0.99353282 9.1007819</span></span>
<span><span class="co">## alcohol_excess 0.58941296 0.23284878 1.5307532</span></span></code></pre>
<p>Since there is some misclassification in the infector/non-infector
status, we can also run a logistic regression with misclassification
using a simple Bayesian method. This is done by setting the
<code>method</code> argument in <code>regression</code> to
“bayesian_logistic_misclass”.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Require that all covariates have columns of type numeric</span></span>
<span><span class="va">numeric_data</span> <span class="op">&lt;-</span> <span class="va">cleaned_data</span></span>
<span><span class="va">numeric_data</span><span class="op">$</span><span class="va">agenew</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">numeric_data</span><span class="op">$</span><span class="va">agenew</span><span class="op">)</span></span>
<span><span class="va">numeric_data</span><span class="op">$</span><span class="va">genderf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">numeric_data</span><span class="op">$</span><span class="va">genderf1</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">numeric_data</span><span class="op">$</span><span class="va">hivfinal_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">numeric_data</span><span class="op">$</span><span class="va">hivfinal_new</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">samples_logistic</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regression.html">regression</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"bayesian_logistic_misclass"</span>,</span>
<span>                               cleaned_data <span class="op">=</span> <span class="va">numeric_data</span>,</span>
<span>                               prob_source <span class="op">=</span> <span class="va">prob_source</span>,</span>
<span>                               prob_cutoff <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                               sensitivity <span class="op">=</span> <span class="fl">0.28</span>,</span>
<span>                               specificity <span class="op">=</span> <span class="fl">0.97</span><span class="op">)</span></span>
<span><span class="co"># Pull the samples for the hiv coefficient, which are the log adjusted</span></span>
<span><span class="co"># odds ratios for the logistic regression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">samples_logistic</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exploration-of-the-pipeline">Exploration of the Pipeline<a class="anchor" aria-label="anchor" href="#exploration-of-the-pipeline"></a>
</h2>
<p>This package makes it very convenient to change parts of the pipeline
and explore the impact of those changes on the results. For example,
suppose we want to change the sampling fraction prior for TransPhylo,
get new results, do linear regression with those results, and compare
them to the old results.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run TransPhylo on the MCC trees with a new sampling fraction prior</span></span>
<span><span class="co"># Using same MCC trees we have already read in</span></span>
<span><span class="va">prob_source_new_pi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_TransPhylo.html">run_TransPhylo</a></span><span class="op">(</span><span class="va">mcc_trees</span>,</span>
<span>                                     type <span class="op">=</span> <span class="st">"mcctrees"</span>,</span>
<span>                                     cluster_dict <span class="op">=</span> <span class="va">all_cluster_assignments</span>,</span>
<span>                                     out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mcc_trees_example"</span>,</span>
<span>                                                           <span class="st">"TransPhylo"</span>,</span>
<span>                                                           package <span class="op">=</span> <span class="st">"beast2tpPipeline"</span><span class="op">)</span>,</span>
<span>                                     output_name <span class="op">=</span> <span class="st">"tp_res_changed_sampling_frac"</span>,</span>
<span>                                     w.shape <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                     w.scale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                                     prior_pi_a <span class="op">=</span> <span class="fl">26</span>,</span>
<span>                                     prior_pi_b <span class="op">=</span> <span class="fl">74</span>,</span>
<span>                                     share <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"neg"</span>, <span class="st">"off.r"</span><span class="op">)</span>,</span>
<span>                                     startNeg <span class="op">=</span> <span class="fl">1.48</span>,</span>
<span>                                     mcmcIterations <span class="op">=</span> <span class="fl">100000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at some traceplots again</span></span>
<span><span class="va">transphylo_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">transphylo_res_file</span><span class="op">)</span></span>
<span><span class="fu">TransPhylo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TransPhylo/man/plotTraces.html" class="external-link">plotTraces</a></span><span class="op">(</span><span class="va">transphylo_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/example_traceplots_pi2.svg"><!-- --></p>
<p>Now we have the updated probabilities, and we can compare their
distribution to their distribution under the old sampling fraction
prior.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">prob_source</span><span class="op">$</span><span class="va">prob</span>, main <span class="op">=</span> <span class="st">"Distribution of TransPhylo Probabilities"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Probability of being an infector"</span>, col <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="mcc_tree_pipeline_files/figure-html/prob_source_histogram_pi2-1.png" width="700"></p>
<p>We can look at individual changes in probabilities via a scatterplot,
where each point represents a sample, the x-axis gives the probability
under the old sampling fraction prior, and the y-axis gives the
probability under the new sampling fraction prior.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prob_source_new_pi</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"prob_source_new_pi"</span><span class="op">)</span></span>
<span><span class="va">prob_source_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">prob_source</span>, <span class="va">prob_source_new_pi</span>, by <span class="op">=</span> <span class="st">"SampleID"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">prob_source_pairs</span><span class="op">$</span><span class="va">prob_source</span>, <span class="va">prob_source_pairs</span><span class="op">$</span><span class="va">prob_source_new_pi</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Prob Source Under Beta(1,19) Sampling Fraction Prior"</span>, ylab <span class="op">=</span> <span class="st">"Prob Source Under Beta(26,74) Sampling Fraction Prior"</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"Comparison of TransPhylo Probabilities of Being an Infection Source"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span></span>
<span><span class="va">spline_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/smooth.spline.html" class="external-link">smooth.spline</a></span><span class="op">(</span><span class="va">prob_source_pairs</span><span class="op">$</span><span class="va">prob_source</span>,</span>
<span>                            <span class="va">prob_source_pairs</span><span class="op">$</span><span class="va">prob_source_new_pi</span>,</span>
<span>                            df <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">spline_fit</span>, col <span class="op">=</span> <span class="st">"darkgreen"</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="mcc_tree_pipeline_files/figure-html/prob_source_scatterplot-1.png" width="700"></p>
<p>We can also redo our logistic (or other) regression with these new
probabilities, and look at the new odds ratios.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prob_source_new_pi</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"prob_source"</span><span class="op">)</span></span>
<span><span class="va">mdl_logistic_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regression.html">regression</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"logistic"</span>,</span>
<span>                               cleaned_data <span class="op">=</span> <span class="va">cleaned_data</span>,</span>
<span>                               prob_source <span class="op">=</span> <span class="va">prob_source_new_pi</span>,</span>
<span>                               prob_cutoff <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mdl_logistic_new</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## glm(formula = tp_source ~ ., family = "binomial", data = cleaned_data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##                Estimate Std. Error z value Pr(&gt;|z|)    </span></span>
<span><span class="co">## (Intercept)     -2.1106     0.5318  -3.969 7.22e-05 ***</span></span>
<span><span class="co">## hivfinal_new     0.3901     0.3611   1.080    0.280    </span></span>
<span><span class="co">## genderf1         0.1623     0.3811   0.426    0.670    </span></span>
<span><span class="co">## tb_everf3       -0.0882     0.4450  -0.198    0.843    </span></span>
<span><span class="co">## smokef1          0.4950     0.4820   1.027    0.304    </span></span>
<span><span class="co">## alcohol_excess  -0.1397     0.4434  -0.315    0.753    </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Dispersion parameter for binomial family taken to be 1)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Null deviance: 213.02  on 239  degrees of freedom</span></span>
<span><span class="co">## Residual deviance: 210.53  on 234  degrees of freedom</span></span>
<span><span class="co">## AIC: 222.53</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Fisher Scoring iterations: 4</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get odds ratios (adjusted for other covariates)</span></span>
<span><span class="va">adj_odds_ratios_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>est <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">mdl_logistic_new</span><span class="op">$</span><span class="va">coefficients</span><span class="op">)</span>,</span>
<span>                                  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mdl_logistic_new</span><span class="op">$</span><span class="va">coefficients</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">adj_odds_ratios_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">adj_odds_ratios_new</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/confint.html" class="external-link">confint</a></span><span class="op">(</span><span class="va">mdl_logistic_new</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Waiting for profiling to be done...</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj_odds_ratios_new</span></span></code></pre></div>
<pre><code><span><span class="co">##                      est      2.5 %    97.5 %</span></span>
<span><span class="co">## (Intercept)    0.1211602 0.03975307 0.3245642</span></span>
<span><span class="co">## hivfinal_new   1.4771205 0.73109189 3.0362124</span></span>
<span><span class="co">## genderf1       1.1761967 0.55457392 2.4899201</span></span>
<span><span class="co">## tb_everf3      0.9155799 0.39564919 2.3125728</span></span>
<span><span class="co">## smokef1        1.6405469 0.65534985 4.4001744</span></span>
<span><span class="co">## alcohol_excess 0.8696210 0.36854637 2.1198239</span></span></code></pre>
<p>And we can compare the odds ratio for HIV status from the two models:
<img src="mcc_tree_pipeline_files/figure-html/compare_odds_ratios-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessalyn Sebastian.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
